### PR TITLE
Demo chat server

### PR DESCRIPTION
**Pull request by ege0zcan**
## Notes
A barebones chat server built using the codewhisperer streaming client. Currently it only supports basic conversations and nothing more. To test it, the vscode test client can be used by  setting the `ENABLE_CHAT_DEMO` flag to `true` in `launch.json` which will spawn a simple chat UI in a webview 

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: server/aws-lsp-codewhisperer/src/language-server/codeWhispererServer.ts [hunk 1]
```diff
@@ -259,18 +259,6 @@ const mergeSuggestionsWithRightContext = (
     })
 }
 
-// Checks if any suggestion in list of suggestions matches with left context of the file
-const hasLeftContextMatch = (suggestions: Suggestion[], leftFileContent: string): boolean => {
```
Comment (ege0zcan) [inline server/aws-lsp-codewhisperer/src/language-server/codeWhispererServer.ts:263]: This change is not related to the chat server but noticed this unused function and removed it

File: client/vscode/src/chatActivation.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    });
+    
+    sendButton.addEventListener('click', sendMessage);
+    inputField.addEventListener('keyup', (event) => {
+        if (event.key === 'Enter') {
+            sendMessage();
+        }
+    });
+    
+    function sendMessage() {
```
Comment (volodkevych) [inline client/vscode/src/chatActivation.ts:115]: I think you'll need to merge this with my changes, sorry.

File: server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                        lsp.sendProgress(chatRequestType, params.partialResultToken, chatResult)
+                    }
+                }
+            }
+
+            return chatResult
+        }
+        chat.onChatPrompt(onChatPromptHandler)
+
+        logging.log('Codewhisperer server has been initialised')
```
Comment (volodkevych) [inline server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts:60]: "Q Chat"?

File: server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            let chatResult: ChatResult = { body: '' }
+
+            for await (const value of response.generateAssistantResponseResponse) {
+                if (token.isCancellationRequested) {
+                    // handle cancellation
+                    return chatResult
+                }
+
+                if (value?.assistantResponseEvent?.content) {
+                    chatResult.body += value?.assistantResponseEvent?.content
```
Comment (volodkevych) [inline server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts:47]: nit: question marks are not needed here

File: server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+                    // If request attached a partial result token, send partial progress response
+                    if (params.partialResultToken) {
+                        lsp.sendProgress(chatRequestType, params.partialResultToken, chatResult)
```
Comment (volodkevych) [inline server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts:51]: For `chatResult`, shouldn't we return just delta here?

Comment (ege0zcan) [reply server/aws-lsp-codewhisperer/src/language-server/qChatServer.ts:51]: No, the partial result should grow and shouldn't be sent as deltas as per the LSP spec. Sending deltas would also cause issues if the partial messages are not received in order

**Pull request closed without merging**